### PR TITLE
Extensible frame content

### PR DESCRIPTION
**Pull request by toots**
This PR adds support for extensible, abstract frame content. It implements the base work to support arbitrary media content, including encoded media!

The idea is to abstract frame content using extensible variant types. The `Frame_content` module contains top-level types for `data`, `format` and `params`. Each of these top-level types represent any possible abstract data, format or params.

A module type, `ContentSpecs` and a module functor, `MkContent` are provided to register new content. Once a new content is registered, the generated module provides methods to convert back and forth from the top-level types to the specific types (`get_*` and `lift_*`).

Content is described by this module:
```ocaml
(** Generic content registration API. *)

(* Raised during any invalid operation below. *)
exception Invalid

module Contents : sig
  type kind
  type format
  type data
end

module type ContentSpecs = sig
  type kind
  type param
  type data

  (** Data *)

  val make : param list -> data
  val blit : data -> int -> data -> int -> int -> unit
  val bytes : data -> int
  val copy : data -> data

  (** Params *)

  val params : data -> param list
  val merge : param list -> param list -> param list
  val string_of_param : param -> string

  (* [param_of_string "label" "value"] *)
  val param_of_string : string -> string -> param option

  (** Kind *)

  val kind : kind
  val default_params : kind -> param list
  val string_of_kind : kind -> string
  val kind_of_string : string -> kind option
end

module type Content = sig
  include ContentSpecs

  (** Data *)

  val is_data : Contents.data -> bool
  val lift_data : data -> Contents.data
  val get_data : Contents.data -> data

  (** Format *)

  val is_format : Contents.format -> bool
  val lift_params : param list -> Contents.format
  val get_params : Contents.format -> param list

  (** Kind *)

  val is_kind : Contents.kind -> bool
  val lift_kind : kind -> Contents.kind
  val get_kind : Contents.kind -> kind
end

module MkContent (C : ContentSpecs) :
  Content
    with type kind = C.kind
     and type param = C.param
     and type data = C.data

type format = Contents.format
type kind = Contents.kind
type data = Contents.data

(** Data *)

val make : format -> data
val blit : data -> int -> data -> int -> int -> unit
val bytes : data -> int
val copy : data -> data

(** Format *)

val format : data -> format
val merge : format -> format -> unit
val string_of_format : format -> string

(* [format_of_param "label" "value"] *)
val format_of_param : string -> string -> format

(** Kind *)

val kind : format -> kind
val default_format : kind -> format
val string_of_kind : kind -> string
val kind_of_string : string -> kind

(** Internal content types. *)

(* None content type is abstract and only used
   via its params and data. *)
module None : sig
  val data : Contents.data
  val format : Contents.format
  val is_format : Contents.format -> bool
end

module Audio : sig
  include
    Content
      with type kind = [ `Pcm ]
       and type param = [ `Mono | `Stereo | `Five_point_one ]
       and type data = Audio.Mono.buffer array

  val channels_of_format : Contents.format -> int
  val format_of_channels : int -> Contents.format
end

module Video : Content with type kind = [ `Yuv420p ] and type data = Video.t

module Midi :
  Content
    with type kind = [ `Midi ]
     and type param = [ `Channels of int ]
     and type data = MIDI.Multitrack.buffer

val default_audio : unit -> Contents.format
val default_video : unit -> Contents.format
val default_midi : unit -> Contents.format
val is_internal : kind -> bool
```

Then, frames and frame kinds become:
```ocaml
(** Operations on frames, which are small portions of streams. *)

(** {2 Frame definitions} *)

type 'a fields = { audio : 'a; video : 'a; midi : 'a }

(** High-level description of the content. *)
type kind =
  [ `Any
  | `Internal
  | `Kind of Frame_content.kind
  | `Format of Frame_content.format ]

(* Some common kind. Audio is PCM, video YUV420p *)
val none : kind
val audio_pcm : kind
val audio_n : int -> kind
val audio_stereo : kind
val audio_mono : kind
val video_yuv420p : kind
val midi_native : kind
val midi_n : int -> kind

type content_kind = kind fields

(** Precise description of the channel types for the current track. *)
type content_type = Frame_content.format fields

type content = Frame_content.data fields
```
The rest of the changes are mostly lifting this new API with some cleanup along the way.

Noteworthy changes:
- Default video content become mono-channel by default
- Audio channels are forced to one of the known channel layouts, i.e. `mono`, `stereo` or `5.1`
- Content description become: `source(audio=pcm(stereo),video=yuv420p,midi=none)`.

### CODE REVIEW

File: src/source.ml [hunk 1]
```diff
@@ -472,38 +436,26 @@ class virtual operator ?(name = "src") kind sources =
...
               (Frame.string_of_content_type ct);
             ctype <- Some ct;
             ct
 
-    method private set_kind =
-      assert (not accessed_kind);
-      List.iter (fun s -> Kind.unify self#kind_var s#kind_var) sources
-
-    initializer self#set_kind
+    method private channels =
```
Comment (smimram) [inline src/source.ml:457]: We might have channels of other things:
```suggestion
    method private audio_channels =
```

File: src/sources/blank.ml [hunk 2]
```diff
@@ -48,25 +48,28 @@ class blank ~kind duration =
...
-      Audio.clear
-        (Audio.sub content.Frame.audio
-           (Frame.audio_of_master position)
-           (Frame.audio_of_master length));
+      ( try
+          Audio.clear
+            (Audio.sub (AFrame.pcm ab)
+               (Frame.audio_of_master position)
+               (Frame.audio_of_master length))
+        with Frame_content.Invalid -> () );
```
Comment (smimram) [inline src/sources/blank.ml:58]: What is the purpose of this catched exception here?

File: src/conversions/drop.ml [hunk 3]
```diff
@@ -20,23 +20,19 @@
...
         ~audio ~video ~midi
         ~converter:(fun ~frame:_ _ -> ())
         source
-
-    method set_kind =
-      let kind = source#kind_var in
-      let kind = if audio then Source.Kind.set_audio kind 0 else kind in
-      let kind = if video then Source.Kind.set_video kind 0 else kind in
-      let kind = if midi then Source.Kind.set_midi kind 0 else kind in
-      Source.Kind.unify self#kind_var kind
```
[the remark this answers is not in the document]
Comment (toots) [reply src/conversions/drop.ml:39]: It is superseded by a new declarative mechanism here: https://github.com/savonet/liquidsoap/pull/1237/files#diff-4150b806c208f4bfe06295160c3ed065R354

File: src/decoder/aac_decoder.ml [hunk 4]
```diff
@@ -157,7 +157,12 @@ let file_type filename =
       in
       log#info "Libfaad recognizes %S as AAC (%dHz,%d channels)." filename rate
         channels;
-      Some { Frame.audio = channels; video = 0; midi = 0 })
+      Some
+        {
+          Frame.audio = Frame_content.Audio.params_of_channels channels;
+          video = Frame_content.None.params;
```
[the remark this answers is not in the document]
Comment (toots) [reply src/decoder/aac_decoder.ml:163]: Agreed. In CPO terms, `params` is top, the fully specified content kind while `format` is a partially specified content kind, so each abstract content needs to have both. I think the structure is:
```
`Any < `Internal < { none, pcm, yuv420p, midi } format < { none, pcm, yuv420p, midi } params
```

File: src/source.ml (hunk 1 above)
Comment (toots) [reply src/source.ml:457]: Word.

File: src/sources/blank.ml (hunk 2 above)
Comment (toots) [reply src/sources/blank.ml:58]: `Frame.pcm` requires `pcm` content type but the `Frame` may contain other type of contents. I decided to stick with only one type of exception coming out of `Frame_content` for now. I can add more but I'm not sure if that'll be useful.

File: src/decoder/aac_decoder.ml (hunk 4 above)
[the remark this answers is not in the document]
Comment (toots) [reply src/decoder/aac_decoder.ml:163]: Also, this is abstracted in the `Frame` module and declared as `Frame.none`. This operator should switch to that value, which would make things more clear.

[the remark this answers is not in the document]
Comment (toots) [reply src/decoder/aac_decoder.ml]: Renamed!

**Comment by toots** [comment]
Ok, I've done some renaming, thanks to @smimram for pointing them out. This PR is ready for me. I'll let it sit until tomorrow and will most likely merge. It'll also bump version to 2.0!! 🎉 

..also had a quick look at the doc and we're gonna have a lot of stuff to change there next..

**Pull request merged**